chore: housekeeping

TTtie opened the pull request:
This PR focuses on housekeeping and implementing what might have been missed after merging everything together.

Pretty much a prerequisite for 0.1.0 to happen :)

TTtie committed:
add pull request CI

TTtie committed:
forgot to checkout the repository :(

TTtie committed:
do not fail fast on CI

TTtie committed:
deprecate the default export function call

TTtie committed:
correct invalid jsdoc

TTtie committed:
add guildAuditLogEntryCreate event

TTtie committed:
Add deprecation notices to examples using deprecated interfaces

TTtie committed:
Add missing deprecation comment to esm.mjs

TTtie committed:
export ForumChannel on the main module

TTtie committed:
Update flags and interfaces related to OAuth application data

This also includes deprecation of undocumented flags

TTtie committed:
fix linting

WilsontheWolf commented:
Hey I was playing around with the guildAuditLogEntryCreate event added in this pr and I found that an error occurs with created invites, when I try accessing entry.target.

Output:
```
ready as testbot
New audit log 1066073242063872020
Type: INVITE_CREATE 40
Reason: null
User: undefined undefined
node:events:491
      throw er; // Unhandled 'error' event
      ^

TypeError: Cannot read properties of undefined (reading 'id')
    at Collection.update (/home/shorty/projects/bot-messing-around/node_modules/@projectdysnomia/dysnomia/lib/util/Collection.js:30:17)
    at new Invite (/home/shorty/projects/bot-messing-around/node_modules/@projectdysnomia/dysnomia/lib/structures/Invite.js:39:70)
    at get target [as target] (/home/shorty/projects/bot-messing-around/node_modules/@projectdysnomia/dysnomia/lib/structures/GuildAuditLogEntry.js:112:20)
    at Client.<anonymous> (file:///home/shorty/projects/bot-messing-around/index.mjs:21:33)
    at Client.emit (node:events:513:28)
    at Shard.emit (/home/shorty/projects/bot-messing-around/node_modules/@projectdysnomia/dysnomia/lib/gateway/Shard.js:229:26)
    at Shard.wsEvent (/home/shorty/projects/bot-messing-around/node_modules/@projectdysnomia/dysnomia/lib/gateway/Shard.js:2336:22)
    at Shard.onPacket (/home/shorty/projects/bot-messing-around/node_modules/@projectdysnomia/dysnomia/lib/gateway/Shard.js:374:26)
    at Shard._onWSMessage (/home/shorty/projects/bot-messing-around/node_modules/@projectdysnomia/dysnomia/lib/gateway/Shard.js:2465:29)
    at WebSocket.emit (node:events:513:28)
Emitted 'error' event on Client instance at:
    at Shard.emit (/home/shorty/projects/bot-messing-around/node_modules/@projectdysnomia/dysnomia/lib/gateway/Shard.js:229:26)
    at Shard._onWSMessage (/home/shorty/projects/bot-messing-around/node_modules/@projectdysnomia/dysnomia/lib/gateway/Shard.js:2468:18)
    at WebSocket.emit (node:events:513:28)
    at Receiver.receiverOnMessage (/home/shorty/projects/bot-messing-around/node_modules/ws/lib/websocket.js:1180:20)
    at Receiver.emit (node:events:513:28)
    at Receiver.dataMessage (/home/shorty/projects/bot-messing-around/node_modules/ws/lib/receiver.js:541:14)
    at Receiver.getData (/home/shorty/projects/bot-messing-around/node_modules/ws/lib/receiver.js:459:17)
    at Receiver.startLoop (/home/shorty/projects/bot-messing-around/node_modules/ws/lib/receiver.js:158:22)
    at Receiver._write (/home/shorty/projects/bot-messing-around/node_modules/ws/lib/receiver.js:84:10)
    at writeOrBuffer (node:internal/streams/writable:392:12)

Node.js v18.12.1
```

Code: 
```js
import { Client, Constants } from '@projectdysnomia/dysnomia';
import fs from 'node:fs';

const token = fs.readFileSync('token', 'utf8');
const client = new Client(token, {
    gateway: {
        intents: ['guilds', 'guildMessages', 'guildModeration'],
    }
});

client.on('ready', () => {
    console.log('ready as', client.user.username);
});

client.on('guildAuditLogEntryCreate', (entry) => {
    console.log('New audit log', entry.id);
    const typeName = Object.entries(Constants.AuditLogActions).find(([, value]) => value === entry.actionType)?.[0];
    console.log('Type:', typeName, entry.actionType);
    console.log('Reason:', entry.reason);
    console.log('User:', entry.user?.username, entry.user?.id);
    console.log('Target:',entry.target ?  entry.target?.constructor.name : '', entry.targetID);
    
});

client.connect();
```
Also another thing I noticed is that the user field is said to always be defined (at least in intellisense) but it actually is undefined in some cases (Specifcally, when I created the invite right off the boot, so I'm probably not cached).

TTtie committed:
fix incorrect invite object construction in GuildAuditLogEntry#target

TTtie commented:
> Hey I was playing around with the guildAuditLogEntryCreate event added in this pr and I found that an error occurs with created invites, when I try accessing entry.target.

That one should be fixed in https://github.com/projectdysnomia/dysnomia/pull/26/commits/13237a1778b426dea2ceda35be89a34092f215dd

> Also another thing I noticed is that the user field is said to always be defined (at least in intellisense) but it actually is undefined in some cases (Specifcally, when I created the invite right off the boot, so I'm probably not cached).

Indeed, it's an issue related to the incompleteness of the cache - the REST call includes the users referenced in the log, the gateway event doesn't; this will be patched appropriately.

TTtie committed:
update audit log-related interfaces to reflect the current API status

TTtie committed:
promote AutoModerationRule to a class; update appropriately

TTtie committed:
fix(Client): whoops, return AutoModerationRule in editAutoModerationRule

TTtie committed:
Add support for Member#flags

ref: https://github.com/discord/discord-api-docs/commit/5a6c4e10858ae736b09d352b67387a096b4adcdb

TTtie committed:
Support GuildChannel#permissions

Available only on channels received on the interaction resolve object

TTtie committed:
docs: VoiceChannel#topic does not exist

TTtie committed:
add GUILD_DIRECTORY into ChannelTypes enum

TTtie committed:
Update the Message interface

TTtie committed:
Document ability to use flags when creating a message

TTtie committed:
Support rateLimitPerUser in createThreadWithMessage

TTtie committed:
correctly sort the Message#thread property

TTtie committed:
deprecate/get rid of undocumented guild features

server discovery and hub types are undocumented

TTtie committed:
correct getGuildPreview return

TTtie committed:
update Guild#toJSON after removal of discovery fields

TTtie committed:
support GuildPreview#stickers

TTtie committed:
Support GuildIntegration#scopes

TTtie committed:
support passing systemChannelFlags to createGuild()

TTtie committed:
support editing Guild#premiumProgressBarEnabled

TTtie committed:
support rtcRegion and videoQualityMode in createChannel()

TTtie committed:
sort endpoints appropriately

TTtie committed:
Support editing the guild's MFA level

TTtie committed:
return the new MFA level directly

TTtie committed:
correctly sort GuildScheduledEvent properties in the doc

TTtie committed:
Deprecate undocumented scheduled event constants

TTtie committed:
Support fetching event data in getInvite()

TTtie committed:
Support associated event data with StageInstance

TTtie committed:
Update docs related to stage instance privacy levels

TTtie committed:
Support sendStartNotification in createStageInstance

TTtie committed:
emit deprecation on getInvite(invID, withCounts, withExpiration)

TTtie committed:
SticketPack banner_asset_id is optional

WilsontheWolf committed:
feat(MessageFlags): add `SUPPRESS_NOTIFICATIONS` message flag (#30)

* Add the new `SUPPRESS_NOTIFICATIONS` message flag

This adds the new `SUPPRESS_NOTIFICATIONS`, as mentioned in https://github.com/discord/discord-api-docs/pull/5910/files

* Update TS definitions

---------

Co-authored-by: TTtie <me@tttie.cz>

TTtie committed:
support creating new threads in a forum via webhooks

TTtie committed:
remove handling of private_channels on READY

it's been an empty array for bots since 2018, not present in the docs anymore

TTtie committed:
add old default forum layout to channel updates

TTtie committed:
include components and remove mentionedBy from messageUpdate

TTtie committed:
emit possibly uncached guilds in guildBan{Add,Remove} events

TTtie committed:
use PossiblyUncached* types consistently

TTtie committed:
feat(GuildIntegration): add scopes to toJSON()

TTtie committed:
document GuildIntegration#guild

TTtie committed:
Support guild integration gateway events

TTtie committed:
Support text-in-stage

TTtie committed:
doc(Client): specify getGuildAuditLog return type

TTtie committed:
quote dev branch install command
# is a comment in some shells
